### DISCUSSION TITLE
OK Alarms

### DISCUSSION THREAD

**Discussion by wernermorgenstern** [cdklabs/cdk-monitoring-constructs#176] [Q&A] [answered]
I have enabled Alarms and onOKAlarms as well.  I do get every so often the OK Alarms, even though no Warning or Critical Alarms were triggered.  Is it possible to only have OK Alarms triggered, after a warning or critical alarm happened?

**Comment by ghost** [comment]
I don't believe we can, at least out of the box.

The implementation simply depends on the [Action#addOkAction](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_cloudwatch.Alarm.html#addwbrokwbractionactions) method, so it should in theory be triggered anytime an alarm has transitioned from breaching into ok state.

Is it seemingly not actually transitioning from breaching when it triggers the messages?

**Comment by wernermorgenstern** [comment]
@echeung-amzn , yes, correct.  It was not breaching.
Here is one of the Slack Messages I get:
Threshold Crossed: 1 out of the last 3 datapoints [12.0 (23/06/22 13:43:00)] was not greater than the threshold (80.0) (minimum 1 datapoint for ALARM -> OK transition).

**Reply by ghost** [reply]
That seems to imply that it _was_ in an alarm state before. Is the problem that the previous alarm state didn't trigger a warning or critical alarm?

**Comment by wernermorgenstern** [comment]
@echeung-amzn , I think I know the issue.  After going into the AWS Console, I see the alarms are configured as "Treat missing data as missing".  And I have through the Construct only the Critical Alarms enabled (not the warnings).  So when the Lambda Function gets used, then the alarm goes from Missing to OK, and that triggers the OK Alarm.

Is there a way through the construct to configure "Treat missing data as good"?  I did not see that anywhere in the documentation.

**Reply by ghost** [reply]
You should be able to declare `treatMissingDataOverride`s, e.g.:

```ts
addFaultCountAlarm: {
  Critical: {
    maxErrorCount: 2,
    datapointsToAlarm: 20,
    treatMissingDataOverride: TreatMissingData.NOT_BREACHING,
  },
},
```

**Comment by wernermorgenstern** [comment, marked as answer]
@echeung-amzn thank you very much.  I must have missed that somewhere in the documentation.
